File: simple-stack-example-sharedelement-fragment/src/main/java/com/example/fragmenttransitions/BaseKey.java
package com.example.fragmenttransitions;

import android.os.Bundle;
import android.os.Parcelable;

/**
 * Created by Owner on 2017. 08. 08..
 */

public abstract class BaseKey
        implements Parcelable {
    public String getFragmentTag() {
        return toString();
    }

    public final BaseFragment newFragment() {
        BaseFragment fragment = createFragment();
        Bundle bundle = fragment.getArguments();
        if(bundle == null) {
            bundle = new Bundle();
        }
        bundle.putParcelable("KEY", this);
        fragment.setArguments(bundle);
        return fragment;
    }

    protected abstract BaseFragment createFragment();
}

File: simple-stack-example-kotlin-community-sample/src/main/java/com/zhuinden/simplestackkotlindaggerexample/BaseKey.kt
package com.zhuinden.simplestackkotlindaggerexample

import android.os.Bundle
import android.os.Parcelable

abstract class BaseKey : Parcelable {
    val fragmentTag: String
        get() = toString()

    fun newFragment(): BaseFragment = createFragment().apply {
        arguments = (arguments ?: Bundle()).also { bundle ->
            bundle.putParcelable("KEY", this@BaseKey)
        }
    }

    protected abstract fun createFragment(): BaseFragment
}